### ISSUE TITLE
New release

### ISSUE THREAD

**Issue by amoghe** [librsync/librsync#107]
The last release was 2.0.0 in Nov 2015. When would it be appropriate to put out a new release with all the new work on master? ( I'm most interested in the speedup for `rdiff` )

**Comment by dbaarda** [comment]
Agree a new release is due. There are a couple of pull requests I'd like resolved before we do this though; #73 #101 and #96, which would fix up platform filesystem problems and make it easier for a new Debian package release.

Additionally #105 would also help improve rdiff, but I consider that a bit optional for a new release. I also have a related pull request on the way that includes heuristics for better default block and strongsum sizes that also significantly speed it up.

**Comment by dbaarda** [comment]
The platform filesystem problems are all resolved by #111.  #105 has gone in too. 

I've created #109 for better heuristics for block and strongsum sizes but it's a bit contentious and involves extending the public API so I don't think it should go in before a release.

This leaves #96 as the only outstanding patch blocking a release. I'll try to get to that soon.

**Comment by wRAR** [comment]
So will #25 be fixed in the new release?

**Comment by amoghe** [comment]
I ran a test of `rdiff` (HEAD at commit: 52070946ca4ab0bf12c178a65358ccbb1a10dd16) v/s whats available via ubuntu repos (via the rdiff pkg, version: `0.9.7`). The file I'm attempting to diff is a tarball of a linux root filesystem (approx 14GiB, containing a blob of 5GiB of binary data in it).

```
root@9df36799e21c:/librsync# pv /tmp/91.tar | ./rdiff delta /tmp/86.sig.new - /tmp/delta.new
13.1GiB 0:20:03 [11.1MiB/s]

root@9df36799e21c:/librsync# pv /tmp/91.tar | rdiff delta /tmp/86.sig.old - /tmp/delta.old
13.1GiB 1:15:58 [2.94MiB/s] 
```

So you can see that the local build is significantly faster. Im curious as to whether #105 would be the primary reason for this, and whether I can achieve any speedups (by tuning the buffer sizes) using the old version (while I wait for the new version to be officially released).

**Comment by dbaarda** [comment]
Regarding #25, I've updated that bug with next steps. Unfortunately I don't have a platform to test this or much time to work on it, so unless someone else does I don't think it'll get fixed before the next release.

Regarding speed improvements, there are quite a lot of changes since 0.9.7 that will have contributed, but #105 is probably one of the smaller improvements. The biggest would probably be the new hash table in #95 and refined in #98. These are significant code changes you can't replicate in 0.9.7 with just flags.

The best way to speed up 0.9.7 is a combination of larger buffer sizes(see #105) and a larger block size (see #109). For signature, you want --block-size=<sqrt(file-size)>, and for delta, you want --input-size=<block-size> --output-size=<10 + 4*block-size>. The larger --block-size will give you slightly larger delta's for a huge performance increase.

**Comment by wRAR** [comment]
I see, in that case the next release still won't be in Debian unfortunately.

**Comment by dbaarda** [comment]
Hmm... in that case lets see if we can fix #25 by just pulling in updated blake2 sources. I really want the next release to end up in Debian... 0.9.7 is really old and it's about time a newer version started propagating out there.

**Comment by Salamek** [comment]
@dbaarda if we are going  to update blake2 src... maybe use git submodule for it this time ? so we dont need to maintain our own copy in repo...

https://stackoverflow.com/questions/7813030/how-can-i-have-linked-dependencies-in-a-git-repo
https://github.com/BLAKE2

so next update would be simple `git pull` in submodule

**Comment by dbaarda** [comment]
@Salamek I'm looking at it right now. Unfortunately the blake2 reference source doesn't #include "config.h" so either needs a tiny modification before we can use it, or we need include-wrappers around it to do it.

After looking at the differences between the blake2 reference implementation and libb2 I'm also a little wary of the upstream sources changing in ways that break us. Right now updating our snapshot to the blake2 reference implementation is the smallest delta so I'm perusing that. The next thing would be to make cmake prefer the local libb2 library if one exists.

**Comment by dbaarda** [comment]
Hmmm... only my Intel(R) Atom(TM) CPU D525   @ 1.80GHz CPU blake2b seems to be faster without NATIVE_LITTLE_ENDIAN defined, even though it's little endian.

I see that the reference implementation now uses memcpy() for NATIVE_LITTLE_ENDIAN, but libb2 uses just a cast but also requires defined(HAVE_ALIGNED_ACCESS_REQUIRED). So I suspect alignment plays a part on whether you want NATIVE_LITTLE_ENDIAN set for the reference implementation.

I have a feeling we probably don't want NATIVE_LITTLE_ENDIAN set ever now for the reference implementation.

Also interesting is the new reference implementation seems to have fixed #103, which makes me optimistic that this new implementation will also fix #25.

**Comment by dbaarda** [comment]
I've just merged #113 which fixed #25 and #103. The only outstanding pull request that I'd like to resolve before a release is #96 to update the Debian packaging scripts.

**Comment by dbaarda** [comment]
I've just merged in a bunch of fixes for minor bugs, and there are no outstanding bugs of any significance worth blocking a release for. The NEWS.md file has been updated to include everything since the last release and is pretty extensive.

The #96 debian packaging scripts and #47 rpm spec file being out of date are of questionable value to fix, and maybe we should just delete them. I don't think it's worth waiting for them to be resolved before doing a release.

I think we should release v2.0.1 right now.
